File: llm/knowledge.py
import os
import uuid
import math
from typing import List, Dict, Optional
from sqlalchemy.orm import Session
from sqlalchemy import text, select, func
from langchain_text_splitters import MarkdownHeaderTextSplitter, RecursiveCharacterTextSplitter
from langchain_google_genai import GoogleGenerativeAIEmbeddings
from pypdf import PdfReader
from server.models import KnowledgeItem
from server.database import SessionLocal

# --- Configuration ---
GOOGLE_API_KEY = os.getenv("GOOGLE_API_KEY")
EMBEDDING_MODEL = "models/gemini-embedding-001"
EMBEDDING_DIM = 768

# We use two separate embedders for optimal performance as per Gemini docs
# 1. RETRIEVAL_DOCUMENT: For ingesting content
# 2. RETRIEVAL_QUERY: For searching context

_doc_embedder = None
_query_embedder = None

def _get_doc_embedder():
    """Returns embedder optimized for document storage."""
    global _doc_embedder
    if _doc_embedder is None:
        key = os.getenv("GOOGLE_API_KEY")
        if not key:
            raise ValueError("GOOGLE_API_KEY is missing.")
        
        _doc_embedder = GoogleGenerativeAIEmbeddings(
            model=EMBEDDING_MODEL,
            google_api_key=key,
            task_type="retrieval_document"
            # We handle dimensionality manually via MRL (slicing)
        )
    return _doc_embedder

def _get_query_embedder():
    """Returns embedder optimized for search queries."""
    global _query_embedder
    if _query_embedder is None:
        key = os.getenv("GOOGLE_API_KEY")
        if not key:
            raise ValueError("GOOGLE_API_KEY is missing.")
            
        _query_embedder = GoogleGenerativeAIEmbeddings(
            model=EMBEDDING_MODEL,
            google_api_key=key,
            task_type="retrieval_query"
        )
    return _query_embedder

def _process_vector(vec: List[float], target_dim: int = 768) -> List[float]:
    """
    Adapts a vector to the target dimension using Matryoshka Representation Learning (MRL).
    1. Slice to target_dim.
    2. Normalize (L2).
    """
    # 1. Slice
    if len(vec) > target_dim:
        vec = vec[:target_dim]
    
    # 2. Normalize
    norm = math.sqrt(sum(x * x for x in vec))
    if norm > 0:
        return [x / norm for x in vec]
    return vec

def _save_splits(splits, organization_id: uuid.UUID, title_prefix: str) -> int:
    """Helper to save splits to DB."""
    db = SessionLocal()
    embedder = _get_doc_embedder()
    try:
        count = 0
        contents = [s.page_content for s in splits]
        if not contents:
            return 0
            
        try:
            # Generate Raw Embeddings
            raw_vectors = embedder.embed_documents(contents)
        except Exception as e:
            print(f"Batch embedding failed: {e}")
            raise e

        for i, split in enumerate(splits):
            content = split.page_content
            # Process Vector (Slice + Normalize)
            vector = _process_vector(raw_vectors[i], EMBEDDING_DIM)
            
            # Metadata handling
            meta_header = ""
            if split.metadata:
                meta_header = " > ".join([str(v) for k, v in split.metadata.items() if k != 'source'])
            
            full_title = f"{title_prefix}"
            if meta_header:
                full_title += f" - {meta_header}"
            
            if not full_title:
                full_title = "General Knowledge"
            
            item = KnowledgeItem(
                organization_id=organization_id,
                title=full_title,
                content=content,
                embedding=vector,
                metadata_=split.metadata
            )
            db.add(item)
            count += 1
        
        db.commit()
        return count
    except Exception as e:
        db.rollback()
        raise e
    finally:
        db.close()

def ingest_knowledge(text_content: str, organization_id: uuid.UUID, title_prefix: str = "") -> int:
    """
    Ingests a Markdown document.
    """
    headers_to_split_on = [
        ("#", "Header 1"),
        ("##", "Header 2"),
        ("###", "Header 3"),
    ]
    splitter = MarkdownHeaderTextSplitter(headers_to_split_on=headers_to_split_on)
    splits = splitter.split_text(text_content)
    
    return _save_splits(splits, organization_id, title_prefix)

def ingest_pdf(file_path: str, organization_id: uuid.UUID, title_prefix: str = "") -> int:
    """
    Ingests a PDF document from a local file path.
    """
    try:
        reader = PdfReader(file_path)
        full_text = ""
        for page in reader.pages:
            extract = page.extract_text()
            if extract:
                full_text += extract + "\n"
            
        # Recursive Splitter for raw text
        splitter = RecursiveCharacterTextSplitter(
            chunk_size=1000,
            chunk_overlap=200,
            separators=["\n\n", "\n", ".", " ", ""]
        )
        splits = splitter.create_documents([full_text])
        
        return _save_splits(splits, organization_id, title_prefix)
    except Exception as e:
        print(f"Error ingesting PDF: {e}")
        raise e

def search_knowledge(
    query: str, 
    organization_id: uuid.UUID, 
    top_k: int = 5,
    vector_threshold: float = 0.65,
    keyword_rank_threshold: int = 5
) -> List[Dict]:
    """
    Performs Hybrid Search using RRF.
    """
    db = SessionLocal()
    embedder = _get_query_embedder()
    try:
        # --- 1. Vector Search ---
        raw_query_vector = embedder.embed_query(query)
        query_vector = _process_vector(raw_query_vector, EMBEDDING_DIM)
        
        vector_results = db.execute(
            select(KnowledgeItem, KnowledgeItem.embedding.cosine_distance(query_vector).label("distance"))
            .filter(KnowledgeItem.organization_id == organization_id)
            .order_by(KnowledgeItem.embedding.cosine_distance(query_vector))
            .limit(top_k)
        ).all()
        
        # --- 2. Keyword Search ---
        keyword_results = db.execute(
            select(KnowledgeItem, func.ts_rank_cd(KnowledgeItem.search_vector, func.websearch_to_tsquery('english', query)).label("rank"))
            .filter(KnowledgeItem.organization_id == organization_id)
            .filter(KnowledgeItem.search_vector.op('@@')(func.websearch_to_tsquery('english', query)))
            .order_by(text("rank DESC"))
            .limit(top_k)
        ).all()
        
        # --- 3. RRF Fusion ---
        candidates = {}
        
        for rank, (item, distance) in enumerate(vector_results, start=1):
            if item.id not in candidates:
                candidates[item.id] = {"item": item, "vec_rank": None, "key_rank": None, "vec_sim": 1 - distance}
            candidates[item.id]["vec_rank"] = rank
            
        for rank, (item, score) in enumerate(keyword_results, start=1):
            if item.id not in candidates:
                candidates[item.id] = {"item": item, "vec_rank": None, "key_rank": None, "vec_sim": 0.0}
            candidates[item.id]["key_rank"] = rank

        rrf_k = 60
        final_results = []
        
        for item_id, data in candidates.items():
            vec_part = 1 / (rrf_k + data["vec_rank"]) if data["vec_rank"] else 0
            key_part = 1 / (rrf_k + data["key_rank"]) if data["key_rank"] else 0
            rrf_score = vec_part + key_part
            
            is_strong_semantic = data["vec_sim"] > vector_threshold
            is_strong_keyword = (data["key_rank"] is not None) and (data["key_rank"] <= keyword_rank_threshold)
            
            if is_strong_semantic or is_strong_keyword:
                final_results.append({
                    "id": str(data["item"].id),
                    "content": data["item"].content,
                    "title": data["item"].title,
                    "score": rrf_score,
                    "reason": "semantic" if is_strong_semantic else "keyword"
                })
        
        final_results.sort(key=lambda x: x["score"], reverse=True)
        return final_results

    except Exception as e:
        print(f"Error in search: {e}")
        raise e
    finally:
        db.close()

File: llm-go/knowledge/knowledge.go
package knowledge

import (
	"database/sql"
	"errors"
	"fmt"
	"math"
	"strings"
	"time"
)

const (
	EmbeddingModel = "models/gemini-embedding-001"
	EmbeddingDim   = 768
)

type EmbeddingProvider interface {
	EmbedDocument(text string) ([]float64, error)
	EmbedQuery(text string) ([]float64, error)
}

type KnowledgeItem struct {
	ID      string
	Title   string
	Content string
	Score   float64
	Reason  string
}

type Service struct {
	DB       *sql.DB
	Embedder EmbeddingProvider
}

func ProcessVector(vec []float64, targetDim int) []float64 {
	if len(vec) > targetDim {
		vec = vec[:targetDim]
	}
	norm := 0.0
	for _, x := range vec {
		norm += x * x
	}
	norm = math.Sqrt(norm)
	if norm == 0 {
		return vec
	}
	out := make([]float64, len(vec))
	for i, x := range vec {
		out[i] = x / norm
	}
	return out
}

func (s *Service) IngestKnowledge(textContent string, organizationID string, titlePrefix string) (int, error) {
	return s.saveSplits(splitMarkdown(textContent), organizationID, titlePrefix)
}

func (s *Service) IngestPDFText(text string, organizationID string, titlePrefix string) (int, error) {
	return s.saveSplits(recursiveSplit(text, 1000, 200), organizationID, titlePrefix)
}

func (s *Service) saveSplits(splits []string, organizationID string, titlePrefix string) (int, error) {
	if s.DB == nil || s.Embedder == nil {
		return 0, errors.New("db and embedder are required")
	}
	count := 0
	for _, content := range splits {
		v, err := s.Embedder.EmbedDocument(content)
		if err != nil {
			return count, err
		}
		vector := ProcessVector(v, EmbeddingDim)
		title := titlePrefix
		if title == "" {
			title = "General Knowledge"
		}
		_, err = s.DB.Exec(`INSERT INTO knowledge_items (id, organization_id, title, content, embedding, metadata)
			VALUES ($1,$2,$3,$4,$5,$6)`, genID(), organizationID, title, content, floatSliceToPGVector(vector), `{}`)
		if err != nil {
			return count, err
		}
		count++
	}
	return count, nil
}

func (s *Service) SearchKnowledge(query string, organizationID string, topK int, vectorThreshold float64, keywordRankThreshold int) ([]KnowledgeItem, error) {
	if s.DB == nil || s.Embedder == nil {
		return nil, errors.New("db and embedder are required")
	}
	qv, err := s.Embedder.EmbedQuery(query)
	if err != nil {
		return nil, err
	}
	qv = ProcessVector(qv, EmbeddingDim)
	vec := floatSliceToPGVector(qv)
	rows, err := s.DB.Query(`
		WITH vector_results AS (
			SELECT id, title, content, 1 - (embedding <=> $1::vector) AS vec_sim,
				row_number() over (order by embedding <=> $1::vector) as vec_rank
			FROM knowledge_items
			WHERE organization_id = $2
			LIMIT $3
		),
		keyword_results AS (
			SELECT id, title, content,
				row_number() over (order by ts_rank_cd(search_vector, websearch_to_tsquery('english', $4)) DESC) as key_rank
			FROM knowledge_items
			WHERE organization_id = $2
			AND search_vector @@ websearch_to_tsquery('english', $4)
			LIMIT $3
		),
		candidates AS (
			SELECT COALESCE(v.id,k.id) id, COALESCE(v.title,k.title) title, COALESCE(v.content,k.content) content,
				COALESCE(v.vec_rank, NULL) vec_rank, COALESCE(k.key_rank, NULL) key_rank,
				COALESCE(v.vec_sim, 0.0) vec_sim
			FROM vector_results v
			FULL OUTER JOIN keyword_results k ON v.id = k.id
		)
		SELECT id, title, content, vec_sim, vec_rank, key_rank,
			(COALESCE(1.0/(60+vec_rank),0) + COALESCE(1.0/(60+key_rank),0)) as rrf_score
		FROM candidates`, vec, organizationID, topK, query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	results := []KnowledgeItem{}
	for rows.Next() {
		var id, title, content string
		var vecSim, rrf float64
		var vecRank, keyRank sql.NullInt64
		if err := rows.Scan(&id, &title, &content, &vecSim, &vecRank, &keyRank, &rrf); err != nil {
			return nil, err
		}
		strongSemantic := vecSim > vectorThreshold
		strongKeyword := keyRank.Valid && int(keyRank.Int64) <= keywordRankThreshold
		if strongSemantic || strongKeyword {
			reason := "keyword"
			if strongSemantic {
				reason = "semantic"
			}
			results = append(results, KnowledgeItem{ID: id, Title: title, Content: content, Score: rrf, Reason: reason})
		}
	}
	return results, nil
}

func splitMarkdown(text string) []string {
	parts := strings.Split(text, "\n\n")
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if p = strings.TrimSpace(p); p != "" {
			out = append(out, p)
		}
	}
	return out
}

func recursiveSplit(text string, size, overlap int) []string {
	if len(text) <= size {
		return []string{text}
	}
	chunks := []string{}
	for start := 0; start < len(text); start += size - overlap {
		end := start + size
		if end > len(text) {
			end = len(text)
		}
		chunks = append(chunks, text[start:end])
		if end == len(text) {
			break
		}
	}
	return chunks
}

func floatSliceToPGVector(v []float64) string {
	parts := make([]string, len(v))
	for i, f := range v {
		parts[i] = fmt.Sprintf("%f", f)
	}
	return "[" + strings.Join(parts, ",") + "]"
}

func genID() string { return fmt.Sprintf("gen-%d", time.Now().UnixNano()) }
